File: mambo-core/mambo-shared/src/main/scala/org/mambo/core/io/Buffer.scala
package org.mambo.core.io

import java.nio.ByteBuffer

/**
 * @author Blackrush
 */
trait Buffer {
  type UByte = Short
  type UShort = Int
  type UInt = Long
  type ULong = BigInt

  var offset: Int

  def data: Array[Byte]
  def rawData: Array[Byte]
  def remaining: Int

  def compact(): Buffer
  def toByteBuffer = ByteBuffer.wrap(rawData)

  def shiftLeft(length: Int)
  def shiftRight(length: Int)

  def readBytes(): Array[Byte]
  def readBytes(length: Int): Array[Byte]
  def writeBytes(bytes: Array[Byte])
  def writeBytes(bytes: Array[Byte], offset: Int, length: Int)

  def writeArray(bytes: Array[Byte])
  def writeArray(bytes: Array[Byte], offset: Int, length: Int)

  def readByte(): Byte
  def writeByte(byte: Byte)

  def readBoolean(): Boolean
  def writeBoolean(boolean: Boolean)

  def readUByte(): UByte
  def writeUByte(ubyte: UByte)

  def readChar(): Char
  def writeChar(char: Char)

  def readShort(): Short
  def writeShort(short: Short)

  def readUShort(): UShort
  def writeUShort(ushort: UShort)

  def readInt(): Int
  def writeInt(int: Int)

  def readUInt(): UInt
  def writeUInt(uint: UInt)

  def readLong(): Long
  def writeLong(long: Long)

  def readULong(): ULong
  def writeULong(ulong: ULong)

  def readFloat(): Float
  def writeFloat(float: Float)

  def readDouble(): Double
  def writeDouble(double: Double)

  def readString(): String
  def readString(length: Int): String
  def writeString(string: String)
}

File: mambo-core/mambo-shared/src/main/java/org/mambo/core/io/Buffer.java
package org.mambo.core.io;

import java.math.BigInteger;
import java.nio.ByteBuffer;

/**
 * @author Blackrush
 */
public interface Buffer {
    int offset();
    void setOffset(int offset);

    byte[] data();
    byte[] rawData();
    int remaining();

    Buffer compact();
    ByteBuffer toByteBuffer();

    void shiftLeft(int length);
    void shiftRight(int length);

    byte[] readBytes();
    byte[] readBytes(int length);
    void writeBytes(byte[] bytes);
    void writeBytes(byte[] bytes, int offset, int length);

    void writeArray(byte[] bytes);
    void writeArray(byte[] bytes, int offset, int length);

    byte readByte();
    void writeByte(byte b);

    boolean readBoolean();
    void writeBoolean(boolean b);

    char readChar();
    void writeChar(char c);

    short readUByte();
    void writeUByte(short ub);

    short readShort();
    void writeShort(short s);

    int readUShort();
    void writeUShort(int us);

    int readInt();
    void writeInt(int i);

    long readUInt();
    void writeUInt(long uint);

    long readLong();
    void writeLong(long ui);

    BigInteger readULong();
    void writeULong(BigInteger ul);

    float readFloat();
    void writeFloat(float f);

    double readDouble();
    void writeDouble(double d);

    String readString();
    String readString(int length);
    void writeString(String s);
}
